dev/bots/analyze.dart

```diff
@@ -0,0 +1,605 @@
+      return;
+    }
+    // Only include files that actually exist, so that we don't try and grep for
+    // nonexistent files, which can occur when files are deleted or moved.
+    final List<String> changedFiles = changedFilesResult.stdout.split('\n').where((String filename) {
+      return new File(filename).existsSync();
+    }).toList();
+    if (changedFiles.isNotEmpty) {
+      await _runCommand('grep',
+        <String>[
+          '--line-number',
+          '--extended-regexp',
+          r'[[:blank:]]$',
+        ] + changedFiles,
+        workingDirectory: flutterRoot,
+        failureMessage: '${red}Whitespace detected at the end of source code lines.$reset\nPlease remove:',
+        expectNonZeroExit: true, // Just means a non-zero exit code is expected.
+        expectedExitCode: 1, // Indicates that zero lines were found.
+      );
+    }
+  }
+}
+
+class EvalResult {
+  EvalResult({
+    this.stdout,
+    this.stderr,
+    this.exitCode = 0,
+  });
+
+  final String stdout;
+  final String stderr;
+  final int exitCode;
+}
+
+Future<EvalResult> _evalCommand(String executable, List<String> arguments, {
+  @required String workingDirectory,
+  Map<String, String> environment,
+  bool skip = false,
+  bool allowNonZeroExit = false,
+}) async {
+  final String commandDescription = '${path.relative(executable, from: workingDirectory)} ${arguments.join(' ')}';
+  final String relativeWorkingDir = path.relative(workingDirectory);
+  if (skip) {
+    _printProgress('SKIPPING', relativeWorkingDir, commandDescription);
+    return null;
+  }
+  _printProgress('RUNNING', relativeWorkingDir, commandDescription);
+
+  final DateTime start = new DateTime.now();
+  final Process process = await Process.start(executable, arguments,
+    workingDirectory: workingDirectory,
+    environment: environment,
+  );
+
+  final Future<List<List<int>>> savedStdout = process.stdout.toList();
+  final Future<List<List<int>>> savedStderr = process.stderr.toList();
+  final int exitCode = await process.exitCode;
+  final EvalResult result = new EvalResult(
+    stdout: utf8.decode((await savedStdout).expand((List<int> ints) => ints).toList()),
+    stderr: utf8.decode((await savedStderr).expand((List<int> ints) => ints).toList()),
+    exitCode: exitCode,
+  );
+
+  print('$clock ELAPSED TIME: $bold${elapsedTime(start)}$reset for $commandDescription in $relativeWorkingDir: ');
+
+  if (exitCode != 0 && !allowNonZeroExit) {
+    stderr.write(result.stderr);
+    print(
+      '$redLine\n'
+      '${bold}ERROR:$red Last command exited with $exitCode.$reset\n'
+      '${bold}Command:$red $commandDescription$reset\n'
+      '${bold}Relative working directory:$red $relativeWorkingDir$reset\n'
+      '$redLine'
+    );
+    exit(1);
+  }
+
+  return result;
+}
+
+String elapsedTime(DateTime start) {
+  return new DateTime.now().difference(start).toString();
+}
+
+Future<Null> _runCommand(String executable, List<String> arguments, {
```

Comment:
There's a lot of repeated code here... Can you please split out the common code and import it into both from there?  Most of the command execution and progress stuff is identical.